flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -101,7 +100,7 @@
 	private final CheckpointStorageLocation targetLocation;
 
 	/** The promise to fulfill once the checkpoint has been completed. */
-	private final CompletableFuture<CompletedCheckpoint> onCompletionPromise;
+	private final CompletableFuture<CheckpointExecutionResult> onCompletionPromise;
```

Comment:
From this PR it does not become obvious for me why we even need `CheckpointExecutionResult`, because the `CompletableFuture` already has the function to represent a succesful case and the exceptional case. So we introduce a new class for a functionality that is already there. We can use it if we have a new `CheckpointExecutionException` that also carries a `CheckpointAbortReason`.

Reply:
@StefanRRichter As the design document mentioned, the `CheckpointExecutionResult ` as a data structure represents the result of checkpoint execution (we split a full checkpoint behavior into two parts: trigger and execute). It will be used in `CheckpointFailureManager` in the next PR(step 2). I know `CompletableFuture` can represent a successful case and an exceptional case. However, in my opinion, use `CheckpointExecutionResult` has two advantage:

* stronger constative
* take concerted action with `CheckpointTriggerResult`

We can just use `onCompletionPromise.complete(new CheckpointExecutionResult(...))` and do not use `onCompletionPromise.completeExceptionally`. It will make the design more clear. What do you think?

Reply:
@yanghua I can read what was written in the design doc about the intented purposes of those classes, but you still could not convince me that they give any additional value. Actually, I would make the same argument against `CheckpointTriggerResult`: why not have a return value for the happy case and a `CheckpointTriggerException` with `CheckpointAbortReason` for the exceptional case on `triggerCheckpoint`? Your advantages are little concrete: 
- about the first: what is good in a class that combines two disjunct outcome cases? I mean except that some of the fields must always be `null`, you need to check boolean methods for the control flow and can trigger ISE if you call something wrong.
- about the second: I don't know what you mean by that. As I see it, the failure manager can later just have 3 different methods like: `reportCheckpointSuccess`, `reportCheckpointTriggertFailure`, `reportCheckpointExecutionFailure` - where the last two receive custom exceptions that bring in the cause enum.

Now I give some more concrete arguments against it:
- More code, more complexity, more bugs (see, we already found some in there).
- We need control flow with many `if`s, instead of just having the normal control flow for the happy case and a clear, separated path for the exceptional case. That is how failure handling is typically done in Java and Flink (I know there are some arguments against it, but that is just how things presently are). 
- People that work with the code need to figure out what disjunct combinations are allowed to be represented by those combined results and can make mistakes. It also somewhat encurages use of `null`.

Reply:
@StefanRRichter Let me give more explanation to my second advantage, `CheckpointTriggerResult` is the existing design in Flink source code currently. I just try to keep the design of `CheckpointExecutionResult` close to `CheckpointTriggerResult`. The implementation of `CheckpointExecutionResult`(`null` and ISE, and provided methods) is similar with `CheckpointTriggerResult`(about `null` and ISE, actually the two `getXXX` methods can be used in combination with two `isXXX` methods, it would be not very serious). In my design document, The `CheckpointFailureManager` just care about `CheckpointTriggerResult` and `CheckpointExecutionResult`, not the three methods you mentioned, they are `handleCheckpointTriggerResult` and `handleCheckpointExecuteResult`. It will make the source code more clear that the failure processing would be split into two stages: **trigger** and **execution**.

Reply:
After my explanation, if you still think the `CheckpointExecutionResult ` is not necessary, I can drop it and accept your suggestion.

Reply:
@yanghua Yes, but take a look at how `CheckpointTriggerResult` is used and then tell me if you don't agree that it would be better to just get an exception as usual. It just checks the boolean and then translates it into `CompletableFuture` anyways. IMO opinion, there is no value in having this class for the code as it currently is. So instead of mimicking and double down on a design that does not look useful, why not correct it?

I also disagree that two methods for the results will make it more clean, just because there are more methods. Inside the methods you have to branch between success and failure case anyways. I think the manager will not care about a succesful trigger, only about the two failure scenarios (trigger, execution) and whether or not there was a checkpoint completed (success).

Reply:
About cleanliness, clean for me is more like: one method - one responsibility. It is a different view on clean.

Reply:
Actually, I agree with you about the advantage of `CompletableFuture`. But about the cleanliness, there also is another design we could combine all the classes about the failure processing. They could be `CheckpointResult`, `CheckpointException`, `CheckpointFailureReason`, `CheckpointFailureManager`. That means we do not emphasize trigger, execution. What do you think?

Reply:
I still think one stage is represented in class name (consider the two keywords `Trigger` and `Decline` in `checkpoint` package). However, another stage is just represented in the code. It is not a good design. I think a good design is that after we view a package of a module, we can learn about its general structure.

Reply:
Basically that sounds very close to what I was suggesting, except that I think we don't need a dedicated `CheckpointResult`. I would imagine this as follows:
- Success case can simply use CompletedCheckpoint or even just the checkpoint id, except if there is a good reason why we need to report more information that we need a bundle class like this.
- Exception cases can report a new type of exception (e.g. `CheckpointException`) with an enum field that carries the reason (e.g. `CheckpointFailureReason`).
- The manager has the corresponding methods to invoke for those cases.

Now I think there are only two questions left for me:
- do we even need `CheckpointResult` or is that unnecessary boilerplate.
- to we only want one type `CheckpointException` or is there a benefit in having two distict exceptions for the trigger case and the execution case. Do you see any advantages to gain from two over one or not?

Reply:
In the manager, we could then also have a mapping from all `CheckpointFailureReason` to some action, e.g. increasing the fail count or not, etc.

Reply:
No I am confused, I see two ways, either we have matching exceptions `Trigger <-> Execution` `Decline <-> Fail`, that is one option. You suggested to merge them, so those words would also go away. So now I was asking if we merge or add the same for the execution path.

Reply:
IMO, we could merge the same execution path. If we follow my thought, we would not distinguish **trigger** and **execution**. I agree with your opinion that we may do not need a dedicated CheckpointResult. What do you think?

Reply:
Sounds good 👍

Reply:
So next, I think I should adjust my design document based on our discussion. Then, I would invite you to review it again and discuss based on the new document. What do you think?

Reply:
Sure, we can do it that way.

Reply:
@StefanRRichter I have refactored my design document, would you please have another look? The whole "T[1]" section reflected our discussion.

Reply:
@StefanRRichter Thanks for your suggestion, I have given more guide about the detail. I described it in "T[1-2]" step 3.

Reply:
I'd like to make sure there are still three steps for T1, T2, and T3. Right? If it is clear, I will start my work for T1 in this PR.

Reply:
Yes, the steps did not change: unified failure types with cause, the manager, then activation and changing the behaviour.

Reply:
Btw, thanks for doing the updates to the design and the implementation efforts :)

Reply:
@StefanRRichter I have refactored my PR based on the new design document. Would you please have a look?